src/Elm/Print.hs

```diff
@@ -206,11 +213,152 @@ elmEnumUniverse t@ElmType{..} = vsep
+    mkCase :: ElmConstructor -> Doc ann
+    mkCase ElmConstructor{..} =
+        conName <+> vars <+> arrow
+            <+> brackets (parens ("tag" <> comma <+> dquotes conName) <> contents)
+      where
+        -- | Constructor name
+        conName :: Doc ann
+        conName = pretty elmConstructorName
+
+        -- | Creates variables: @x1@ to @xN@, where N is the number of the constructor fields.
+        fields :: [Doc ann]
+        fields = take (length elmConstructorFields) $
+            map (pretty . mkText "x") [1..]
+
+        contents :: Doc ann
+        contents = "," <+> parens (dquotes "contents" <> comma <+> "E.list" <+> brackets fieldEncs)
+
+        -- | @encoderA x1@
+        fieldEncs :: Doc ann
+        fieldEncs = concatWith (surround ", ") $
+            zipWith (<+>) (map typeRefEncoder elmConstructorFields) fields
+
+        -- | Makes variable like: @x11@ etc.
+        mkText :: Text -> Int -> Text
+        mkText x i = x <> T.pack (show i)
+
+        vars :: Doc ann
+        vars =  concatWith (surround " ") fields
+
+
+aliasEncoderDoc :: ElmAlias -> Doc ann
+aliasEncoderDoc ElmAlias{..} =
+    encoderDef elmAliasName []
+    <> line
+    <> nest 4 (vsep $ (encoderName elmAliasName <+> "x" <+> equals <+> "E.object")
+            : fieldsEncode elmAliasFields)
+  where
+    fieldsEncode :: NonEmpty ElmRecordField -> [Doc ann]
+    fieldsEncode (fstR :| rest) =
+        lbracket <+> recordFieldDoc fstR
+      : map ((comma <+>) . recordFieldDoc) rest
+     ++ [rbracket]
+
+    recordFieldDoc :: ElmRecordField -> Doc ann
+    recordFieldDoc ElmRecordField{..} = parens $
+            dquotes (pretty elmRecordFieldName)
+         <> comma
+        <+> typeRefEncoder elmRecordFieldType
+        <+> "x." <> pretty elmRecordFieldName
+
+-- | The definition of the @encodeTYPENAME@ function.
+encoderDef
+    :: Text  -- ^ Type name
+    -> [Text] -- ^ List of type variables
+    -> Doc ann
+encoderDef typeName vars =
+    encoderName typeName <+> colon <+> pretty typeName <> typeVarsDoc <+> arrow <+> "Value"
+  where
+    typeVarsDoc :: Doc ann
+    typeVarsDoc = concatWith (surround " ") $ map pretty vars
+
+-- | Create the name of the encoder function.
+encoderName :: Text -> Doc ann
+encoderName typeName = "encode" <> pretty typeName
+
+-- | Converts the reference to the existing type to the corresponding encoder.
+typeRefEncoder :: TypeRef -> Doc ann
+typeRefEncoder (RefPrim elmPrim) = case elmPrim of
+    ElmUnit       -> "const E.null"
+    ElmNever      -> "never"
+    ElmBool       -> "E.bool"
+    ElmChar       -> parens "E.string << String.fromChar"
+    ElmInt        -> "E.int"
+    ElmFloat      -> "E.float"
+    ElmString     -> "E.string"
+    ElmMaybe t    -> parens $ "encodeMaybe" <+> typeRefEncoder t
+    ElmResult l r -> parens $ "encodeEither" <+> typeRefEncoder l <+> typeRefEncoder r
+    ElmPair a b   -> parens $ "encodePair" <+> typeRefEncoder a <+> typeRefEncoder b
```

Comment:
You raised a valid point during discussion about the names for these functions. I think names could be one of the following:

1. `elmStreetEncodeMaybe`
2. `defEncodeMaybe`
3. `elmEncodeMaybe`
4. `maybeEncoder`
5. `maybeEncode`

Or probably something else. What do you think about the names above?

Reply:
I like the first one the most. I guess with this name it's 0 chance of having the troubles with the names collisions.